packages/peregrine/lib/talons/CartPage/GiftOptions/useGiftOptions.js

```diff
@@ -0,0 +1,126 @@
+import { useState, useCallback, useEffect, useMemo } from 'react';
+import throttle from 'lodash.throttle';
+import { useMutation, useLazyQuery } from '@apollo/react-hooks';
+
+import { useCartContext } from '@magento/peregrine/lib/context/cart';
+
+const useGiftOptions = ({ getGiftOptionsQuery, saveGiftOptionsQuery }) => {
+    const [includeGiftReceipt, setIncludeGiftReceipt] = useState(false);
+    const [includePrintedCard, setIncludePrintedCard] = useState(false);
+    const [giftMessage, setGiftMessage] = useState('');
+
+    const [{ cartId }] = useCartContext();
+
+    const [fetchGiftOptions, { data }] = useLazyQuery(getGiftOptionsQuery, {
+        fetchPolicy: 'no-cache'
+    });
+
+    /**
+     * Fetch gift options for a given cart id.
+     */
+    useEffect(() => {
+        if (cartId) {
+            fetchGiftOptions({
+                variables: {
+                    cart_id: cartId
+                }
+            });
+        }
+    }, [cartId, fetchGiftOptions]);
+
+    const [setGiftOptions] = useMutation(saveGiftOptionsQuery);
+
+    const updateGiftOptions = useCallback(
+        optionsToUpdate => {
+            const newGiftOptions = {
+                cart_id: cartId,
+                include_gift_receipt: includeGiftReceipt,
+                include_printed_card: includePrintedCard,
+                gift_message: giftMessage,
+                ...optionsToUpdate
+            };
+            setGiftOptions({
+                variables: newGiftOptions
+            });
+        },
+        [
+            setGiftOptions,
+            cartId,
+            includeGiftReceipt,
+            includePrintedCard,
+            giftMessage
+        ]
+    );
+
+    /**
+     * Throttling message update. Only make 1 mutation
+     * every 5 seconds. This is to save on bandwidth.
+     *
+     * More info: https://lodash.com/docs/4.17.15#throttle
+     */
+    const throttledMessageUpdate = useMemo(() => {
+        return throttle(
+            (updateGiftOptions, newGiftMessage) => {
+                updateGiftOptions({
+                    gift_message: newGiftMessage
+                });
+            },
+            5000,
```

Comment:
5 seconds is pretty long. Since this no longer depends on a network request and is now just saving to cache maybe we can reduce this to half a second?

Reply:
The intent behind 5 seconds and other decisions I took while writing this hook were to save on dev time once the gift options support from the GQL team is available. The plan was to just delete the local resolvers and remove `@client` directive.

`useGiftOptions` should not be touched once gift options are available through GQL.